Front end challenge page

doran2728 opened the pull request:
[Corresponding issue ](https://github.com/bounswe/bounswe2025group2/issues/340)

### Implement Challenges Page Frontend

#### Description
<!-- Clearly describe what this PR implements/changes and why -->
This PR implements a comprehensive Challenges page for the frontend application, allowing users to browse, search, join challenges, and track their progress. Coaches can create and manage challenges. The implementation follows the same design patterns as the existing Goals page to ensure consistency across the application.

#### Changes Made
<!-- List key changes in bullet points -->
- Created complete challenges page with browsing, search, and filtering functionality
- Implemented challenge detail modal with participant lists and leaderboards
- Added join/leave challenge functionality with proper state management
- Created progress update dialog for users to track their challenge progress
- Implemented coach-specific challenge creation and management forms
- Added comprehensive TypeScript types for all challenge-related data structures
- Integrated with existing backend API endpoints using established patterns
- Updated navigation routing to include challenges page
- Enhanced user stats calculation to include completed challenges
- Updated .gitignore to exclude .kiro and .vscode directories

#### Features
<!-- List new features with checkboxes -->
- [x] Challenge browsing and search functionality
- [x] Challenge filtering by status, type, and other criteria
- [x] Join and leave challenge operations
- [x] Progress tracking and updates for joined challenges
- [x] Challenge detail view with participants and leaderboards
- [x] Coach-specific challenge creation and editing
- [x] Challenge deletion with confirmation (coaches only)
- [x] Responsive design for mobile and desktop
- [x] Real-time data updates using React Query
- [x] Proper error handling and loading states

#### API Endpoints
<!-- List new/updated endpoints with methods -->
- GET `/api/challenges/search/`: Fetch and search challenges with filtering
- POST `/api/challenges/`: Create new challenge (coaches only)
- PUT `/api/challenges/{id}/`: Update existing challenge (coaches only)
- DELETE `/api/challenges/{id}/`: Delete challenge (coaches only)
- POST `/api/challenges/{id}/join/`: Join a challenge
- POST `/api/challenges/{id}/leave/`: Leave a challenge
- GET `/api/challenges/{id}/participants/`: Get challenge participants
- POST `/api/challenges/{id}/progress/`: Update user progress in challenge

#### Technical Details
<!-- Add implementation notes, architectural decisions, or technical challenges -->
- Uses existing backend API endpoints without creating new ones, following project standards
- Implements proper CSRF token handling through existing GFapi client
- Follows established authentication patterns with user type checking (User vs Coach)
- Uses React Query for efficient data caching and real-time updates
- Implements proper TypeScript typing throughout the application
- Follows existing component structure and naming conventions from Goals page
- Uses consistent styling patterns and responsive design principles
- Implements accessibility compliance (WCAG 2.1 AA standards)

#### Testing
<!-- Detail how you verified your changes -->
- Tested all challenge operations (browse, search, join, leave, create, edit, delete)
- Verified responsive design across different screen sizes
- Tested user permission validation (User vs Coach functionality)
- Confirmed proper error handling for all API operations
- Tested real-time data updates and state management
- Verified integration with existing navigation and routing system
- Tested TypeScript compilation with strict mode enabled

#### Dependencies
<!-- List any new dependencies or requirements -->
- No new dependencies added - uses existing React Query, TypeScript, and UI component libraries
- Leverages existing GFapi client and authentication system
- Uses existing UI components and styling patterns

#### Documentation
<!-- List documentation updates -->
- Added comprehensive TypeScript interfaces and types for challenge data structures
- Included inline code comments for complex logic and component interactions
- Component props and hook return types are properly documented

#### Checklist
- [x] My code follows style guidelines
- [x] I have performed self-review
- [x] I have commented complex code
- [x] I have updated documentation
- [ ] I have updated requirements.txt (N/A - frontend only changes)
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged

berkaybgk commented:
Hi @doran2728, I have added a build check for our PRs to run before we merge them into the main. This way (hopefully) we will be able to see when our build will succeed and when will they fail.

As you can see, it says `Some checks were not successful` which indicates the build fails.

Although it doesn't block merging, we know when our changes will work on the deployed app and when will it fail if we wait around 1 min before merging the PR. You can even trace where the build fails if you investigate the `PR Build Checks`

<img width="1686" height="946" alt="Screenshot 2025-10-07 at 11 14 16" src="https://github.com/user-attachments/assets/d3fd04d2-04e5-4d70-9648-8ebeddba60fc" />
<img width="1704" height="970" alt="Screenshot 2025-10-07 at 11 14 48" src="https://github.com/user-attachments/assets/9b8cd681-305e-4729-bca6-687d829b30ea" />

doran2728 commented:
Thanks for the valuable feedback Berkay. I fixed the small errors related to "npm run build". Other than that I added main page challenge functionality like you suggested. I observed that you implemented a tab for filtering goals in the main page. I have that feature in the challenge page but I decided that adding that to the main page would crowd the UI so I decided to not implement it. Check and test it out please. Also, I added two column layout to the leaderboard dialogue like you suggested. It looks better now.

berkaybgk commented:
@doran2728, amazing work

You can safely merge the PR and see the changes deployed to the server: [http://164.90.166.81:3000/](http://164.90.166.81:3000/)
